### PR TITLE
fix: typetracer `nplike.repeat` & `ak.with_field`

### LINKED ISSUE #2428: Assigning single numbers to typetracers fails
### Version of Awkward Array

2.1.3

### Description and code to reproduce

The following script:
```python3
import json
import awkward as ak
import numpy as np

fromjson = {
    "class": "RecordArray",
    "fields": ["muon"],
    "contents": [
        {
            "class": "ListOffsetArray",
            "offsets": "i64",
            "content": {
                "class": "RecordArray",
                "fields": ["pt", "eta", "phi", "crossref", "thing1"],
                "contents": [
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_pt!",
                    },
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_eta!",
                    },
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_phi!",
                    },
                    {
                        "class": "ListOffsetArray",
                        "offsets": "i64",
                        "content": {
                            "class": "NumpyArray",
                            "primitive": "int64",
                            "inner_shape": [],
                            "parameters": {},
                            "form_key": "muon_crossref_content!",
                        },
                        "parameters": {},
                        "form_key": "muon_crossref_index!",
                    },
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_thing1!",
                    },
                ],
                "parameters": {},
                "form_key": "muon_record!",
            },
            "parameters": {},
            "form_key": "muon_list_outer!",
        },
    ],
    "parameters": {},
    "form_key": "outer!",
}

form = ak.forms.from_json(json.dumps(fromjson))

ttlayout, report = ak._nplikes.typetracer.typetracer_with_report(form, forget_length=True)

ttarray = ak.Array(ttlayout, behavior=ak.behavior)

muons = ttarray.muon

# both of the following fail individually and are required
muons = ak.with_field(ttarray.muon, 0.4, "something")

muons["somethingelse"] = 0.5
```

Fails with a NotImplemented error. This is needed for convenient shorthands to work in dask_awkward:

```
Traceback (most recent call last):
  File "/Users/lgray/coffea-dev/awkward/src/awkward/highlevel.py", line 1027, in __setitem__
    self._layout = ak.operations.with_field(
  File "/Users/lgray/coffea-dev/awkward/src/awkward/operations/ak_with_field.py", line 47, in with_field
    return _impl(array, what, where, highlevel, behavior)
  File "/Users/lgray/coffea-dev/awkward/src/awkward/operations/ak_with_field.py", line 150, in _impl
    out = ak._broadcasting.broadcast_and_apply(
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_broadcasting.py", line 1062, in broadcast_and_apply
    out = apply_step(
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_broadcasting.py", line 1041, in apply_step
    return continuation()
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_broadcasting.py", line 1014, in continuation
    return broadcast_any_list()
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_broadcasting.py", line 563, in broadcast_any_list
    outcontent = apply_step(
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_broadcasting.py", line 1041, in apply_step
    return continuation()
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_broadcasting.py", line 1014, in continuation
    return broadcast_any_list()
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_broadcasting.py", line 609, in broadcast_any_list
    outcontent = apply_step(
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_broadcasting.py", line 1027, in apply_step
    result = action(
  File "/Users/lgray/coffea-dev/awkward/src/awkward/operations/ak_with_field.py", line 127, in action
    backend.nplike.repeat(what, len(base))
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_nplikes/typetracer.py", line 1121, in repeat
    raise NotImplementedError
NotImplementedError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "form_madness10.py", line 79, in <module>
    muons["somethingelse"] = 0.5
  File "/Users/lgray/coffea-dev/awkward/src/awkward/highlevel.py", line 1030, in __setitem__
    self._numbaview = None
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_errors.py", line 56, in __exit__
    self.handle_exception(exception_type, exception_value)
  File "/Users/lgray/coffea-dev/awkward/src/awkward/_errors.py", line 66, in handle_exception
    raise self.decorate_exception(cls, exception)
NotImplementedError: 

See if this has been reported at https://github.com/scikit-hep/awkward/issues
```

### PR DESCRIPTION
**Pull request by agoose77**
This PR fixes a set of miscellaneous typetracer bugs. It doesn't implement a full `repeat` implementation, but it's a start.

### CODE REVIEW

**Comment by lgray** [comment]
This variant also fails, with the PR thus far:
```python3
import json
import awkward as ak
import numpy as np

fromjson = {
    "class": "RecordArray",
    "fields": ["muon"],
    "contents": [
        {
#            "class": "ListOffsetArray",
#            "offsets": "i64",
#            "content": {
                "class": "RecordArray",
                "fields": ["pt", "eta", "phi", "crossref", "thing1"],
                "contents": [
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_pt!",
                    },
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_eta!",
                    },
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_phi!",
                    },
                    {
                        "class": "ListOffsetArray",
                        "offsets": "i64",
                        "content": {
                            "class": "NumpyArray",
                            "primitive": "int64",
                            "inner_shape": [],
                            "parameters": {},
                            "form_key": "muon_crossref_content!",
                        },
                        "parameters": {},
                        "form_key": "muon_crossref_index!",
                    },
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_thing1!",
                    },
                ],
                "parameters": {"__record__": "TestBehavior"},
                "form_key": "muon_record!",
            },
#            "parameters": {},
#            "form_key": "muon_list_outer!",
#        },
    ],
    "parameters": {},
    "form_key": "outer!",
}

@ak.mixin_class(ak.behavior)
class TestBehavior:
    @ak.mixin_class_method(np.add, {"TestBehavior"})
    def add(self, other):
        return ak.zip(
            {
                "pt": self.pt + other.pt,
            },
            with_name="TestBehavior",
            behavior=self.behavior,
        )

form = ak.forms.from_json(json.dumps(fromjson))

ttlayout, report = ak._nplikes.typetracer.typetracer_with_report(form, forget_length=True)

ttarray = ak.Array(ttlayout, behavior=ak.behavior)

muons = ttarray.muon

# both of the following fail individually and are required
muons = ak.with_field(muons, 0.4, "something")

ttarray["othermuons"] = muons

muons["somethingelse"] = 0.5
```

**Comment by agoose77** [comment]
@lgray apologies, my previous commit never pushed to the remote. I've made another commit to address your reproducer (thanks!), does it work now?

File: src/awkward/_nplikes/typetracer.py [hunk 1]
```diff
@@ -1118,7 +1118,25 @@ def repeat(
...
+                size = size * repeats
+            return TypeTracerArray._new(x.dtype, (size,))
+        else:
+            shape = list(x.shape)
+            if isinstance(repeats, TypeTracerArray) and repeats.ndim > 0:
+                raise NotImplementedError
+            elif is_unknown_scalar(repeats):
+                shape[axis] = shape[axis] * self.index_as_shape_item(repeats)
+            else:
+                shape[axis] = shape[axis] * repeats
```
Comment (agoose77) [inline src/awkward/_nplikes/typetracer.py]:
```suggestion
            else:
                size = size * self.index_as_shape_item(repeats)
            return TypeTracerArray._new(x.dtype, (size,))
        else:
            shape = list(x.shape)
            if isinstance(repeats, TypeTracerArray) and repeats.ndim > 0:
                raise NotImplementedError
            else:
                shape[axis] = shape[axis] * self.index_as_shape_item(repeats)
```

**Review by jpivarski — APPROVED** [review]
`x` → `asarray(x)` and `len(x)` → `x.length` are all clearly oversights from previous campaigns.

A "less incomplete" `TypeTracer.repeat` implementation is also an improvement!

Is this the last PR to go before the release? Scanning over the list, it looks like it. (#2425 could go, too, but it doesn't have to.)

**Pull request merged**